# Exercise 4.5 Guided Practice

In [1]:
#import libraries
import pandas as pd
import numpy as np
import os

In [2]:
# create practice df
df_test = pd.DataFrame()

In [3]:
# create a mix type column
df_test['mix'] = ['a', 'b', 1, True]

In [4]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [5]:
# check for mix type columns
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [6]:
#change mix type to string
df_test['mix'] = df_test['mix'].astype('str')

In [7]:
#import df_prods
path = r'C:\Users\Admin\Documents\Data Analytics Course\Achievement 4\08-2020 Instacart Basket Analysis'
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'))

In [8]:
#counting number of missing values
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [9]:
#create subset of missing values
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [10]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [11]:
#number of rows in df before missing rows removed
df_prods.shape

(49693, 5)

In [12]:
# new dataframe with missing rows removed
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [13]:
df_prods_clean.shape

(49677, 5)

In [14]:
# create dataframe from duplicate rows
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [15]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [16]:
#new df without duplicates
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

# Exercise 4.5 Task

## Question 2

In [17]:
# run the df.describe on df_prods and share if anything looks off
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

The maximum for the prices column is odd. It makes me think that 99999 is inputted for missing data points.


## Question 3

In [30]:
#importing df_ords
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col=[0])

In [31]:
df_ords.head()

order_id  customer_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329            1             1                   2                  8   
1   2398795            1             2                   3                  7   
2    473747            1             3                   3                 12   
3   2254736            1             4                   4                  7   
4    431534            1             5                   4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [32]:
#Check for mixed type data in df_ords
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

## Question 4

There is no mixed type data in df_ords.

## Question 5

In [33]:
# count missing values
df_ords.isnull().sum()

order_id                       0
customer_id                    0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

The missing values in the days_since_prior_order column are likely due to first-time customers.

## Question 6

In [34]:
# create new dataframe with a new column "first_time_customer"

df_ords_clean = df_ords
df_ords_clean['new_customer'] = df_ords['days_since_prior_order'].isnull() == True

In [35]:
df_ords_clean

order_id  customer_id  order_number  orders_day_of_week  \
0         2539329            1             1                   2   
1         2398795            1             2                   3   
2          473747            1             3                   3   
3         2254736            1             4                   4   
4          431534            1             5                   4   
...           ...          ...           ...                 ...   
3421078   2266710       206209            10                   5   
3421079   1854736       206209            11                   4   
3421080    626363       206209            12                   1   
3421081   2977660       206209            13                   1   
3421082    272231       206209            14                   6   

         order_hour_of_day  days_since_prior_order  new_customer  
0                        8                     NaN          True  
1                        7                    15.0         False  
2                       12                    21.0         False  
3                        7                    29.0         False  
4                       15                    28.0         False  
...                    ...                     ...           ...  
3421078                 18                    29.0         False  
3421079                 10                    30.0         False  
3421080                 12                    18.0         False  
3421081                 12                     7.0         False  
3421082                 14                    30.0         False  

[3421083 rows x 7 columns]

## Question 7

In [36]:
#checking for duplicates in df_ords_new
df_ords_dups = df_ords_clean[df_ords_clean.duplicated()]

In [37]:
df_ords_dups

Empty DataFrame
Columns: [order_id, customer_id, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order, new_customer]
Index: []

No duplicates found.

## Question 9

In [38]:
# exporting cleaned df_prods and df_ords
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_cleaned.csv'))
df_ords_clean.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_cleaned.csv'))

In [39]:
df_ords_clean

order_id  customer_id  order_number  orders_day_of_week  \
0         2539329            1             1                   2   
1         2398795            1             2                   3   
2          473747            1             3                   3   
3         2254736            1             4                   4   
4          431534            1             5                   4   
...           ...          ...           ...                 ...   
3421078   2266710       206209            10                   5   
3421079   1854736       206209            11                   4   
3421080    626363       206209            12                   1   
3421081   2977660       206209            13                   1   
3421082    272231       206209            14                   6   

         order_hour_of_day  days_since_prior_order  new_customer  
0                        8                     NaN          True  
1                        7                    15.0         False  
2                       12                    21.0         False  
3                        7                    29.0         False  
4                       15                    28.0         False  
...                    ...                     ...           ...  
3421078                 18                    29.0         False  
3421079                 10                    30.0         False  
3421080                 12                    18.0         False  
3421081                 12                     7.0         False  
3421082                 14                    30.0         False  

[3421083 rows x 7 columns]